src/frontend/js/utils/api/joanie.ts

```diff
@@ -0,0 +1,273 @@
+/**
+ *
+ * Joanie API Interface
+ *
+ * Interface to interact with Joanie Api.
+ *
+ * ⚠︎ You should not import this API interface directly from a component,
+ *    instead you should use the `useJoanieApi` Hook.
+ *
+ */
+import { stringify } from 'query-string';
+import type * as Joanie from 'types/Joanie';
+import context from 'utils/context';
+import { AuthenticationApi } from 'utils/api/authentication';
+import { HttpError } from 'utils/errors/HttpError';
+
+interface CheckStatusOptions {
+  fallbackValue: any;
+  ignoredErrorStatus: number[];
+}
+/*
+  An util to manage Joanie API responses.
+  It parses properly the response according to its `Content-Type`
+  otherwise it throws an `HttpError`.
+
+  `options` arguments accept an array of ignoredErrorStatus. If the response
+  fails with one of this status code, the `fallbackValue` will return and no error will
+  be raised.
+*/
+export function checkStatus(
+  response: Response,
+  options: CheckStatusOptions = { fallbackValue: null, ignoredErrorStatus: [] },
+): Promise<any> {
+  if (response.ok) {
+    if (response.headers.get('Content-Type') === 'application/json') {
+      return response.json();
+    }
+    if (response.headers.get('Content-Type') === 'application/pdf') {
+      return response.blob();
+    }
+    return response.text();
+  }
+  if (options.ignoredErrorStatus.includes(response.status)) {
+    return Promise.resolve(options.fallbackValue);
+  }
+
+  throw new HttpError(response.status, response.statusText);
+}
+
+/*
+  Generate default headers used for most of Joanie requests. It defined `Content-Type`
+  to `application/json` and the `Accept-Language` to the active language.
+*/
+function getDefaultHeaders(): Record<string, string> {
+  const $html = document.querySelector('html');
+  const language = $html?.getAttribute('lang') || 'en-US';
+
+  return {
+    'Content-Type': 'application/json',
+    'Accept-Language': language,
+  };
+}
+
+/*
+  Retrieve the access token from `AuthenticationApi` interface then bind it to
+  the request through the `Authorization` header.
+*/
+function fetchWithJWT(routes: RequestInfo, options: RequestInit = {}) {
+  try {
+    const accessToken = AuthenticationApi!.accessToken!();
+    const headers = (options.headers as Record<string, string>) || getDefaultHeaders();
+
+    if (accessToken) {
+      // eslint-disable-next-line @typescript-eslint/dot-notation
+      headers['Authorization'] = `Bearer ${accessToken}`;
+    }
+
+    options.headers = headers;
+
+    return fetch(routes, options);
+  } catch (error) {
+    throw new Error(
+      `Joanie requires JWT Token to fetch data, but the configured authentication
+      api does not contains a method \`accessToken\` to retrieve this information.`,
+    );
+  }
+}
+
+/**
+ * Build Joanie API Routes interface.
+ */
+const getRoutes = (joanieEndpoint: string) => ({
+  user: {
+    creditCards: {
+      get: joanieEndpoint.concat('/api/credit-cards/:id/'),
```

Comment:
Why not using string interpolation here ? `${joanieEndpoint}/api/credit-cards/:id/` ?

Reply:
Because this day, I remembered the `concat` method 😅. I'll use a template string.